### ISSUE TITLE
gRPC Python server not destroyed during shutdown process

### ISSUE THREAD

**Issue by lidizheng** [grpc/grpc#17258]
While implementing Channelz, I found there was a potential leak of C-Core server object.
The Python shutdown process doesn't destroy the C-Core server object in the shutdown process. That means the servers spawned by gRPC Python will always left something in C-Core even if it is shut-downed.

To destroy a server, the comment says:
> Destroy a server.
> Shutdown must have completed beforehand (i.e. all tags generated by
> grpc_server_shutdown_and_notify must have been received, and at least
> one call to grpc_server_shutdown_and_notify must have been made).

So, the desired behavior of shutdown should be
1. CALL `grpc_server_shutdown_and_notify`
2. Python receive the shutdown tag and perform clean-up
3. CALL `grpc_server_destroy`

But in current behavior, the step 3 never performed in the shutdown process.

**Comment by lidizheng** [comment]
Interestingly, the destruction process is weird... I failed to find a way to inject the `grpc_server_destroy`, but I did find a way to make the garbage collect behave correctly.

1. There is a difference of thread termination between Python 2 and Python 3, see https://github.com/grpc/grpc/issues/17214.
2. Direct calling to `grpc_server_destroy` after the grace shutdown and cancel all the calls will result in `SIGABRT`. It is because the serving thread might not terminated at that point. It will crash if the `grpc_server_destroy` called faster than the serving thread termination.
3. The serving thread terminated doesn't mean the destruction of Cython object is triggered. Since the Python `_Server` object owns a reference to the Cython `Server` object, the Cython object will not be garbage collected before the Python object is collected. It seems just an order difference, but in practice the trigger of Cython destruction function will be delaied or just not called.

Thus, in the PR #17259, I explicitly called the `del self._state` to invoke garbage collection. In this way, the C-Core server will be properly destroyed during the Python destruction process.

**Comment by lidizheng** [comment]
Interesting fact about Python 2 gc:
1. The `__del__` will be called multiple time since the reference exist in multiple cycle ([read more](https://stackoverflow.com/questions/18971451/how-many-times-can-del-be-called-per-object-in-python))
1. The class member variable will not be freed until all reference of the class itself got freed, even if the `__del__` of the class is called.

In the case of a unit test tear down with 2 gRPC server, the procedure goes like:
1. Unit test tear down starts
1. First server deletion called
1. First server shutdown
1. Second server deletion called
1. Second server shutdown
1. Unit test tear down finishes
1. (Running other unit tests)
1. (All unit tests ends)
1. Another round of gc starts
1. Second Cython server destroyed
1. First Cython server destroyed

Raw log of garbage collection in 2.7.15:
```
UNIT TEST TEAR DOWN
PYTHON SERVER DELETION <grpc._server._Server object at 0x1078ab810> <_cython.cygrpc.Server object at 0x1078ae0d8>
CANCEL ALL CALLS
NOTIFY SHUTDOWN COMPLETE <_cython.cygrpc.Server object at 0x1078ae0d8>
PYTHON SERVER DELETION DONE <grpc._server._Server object at 0x1078ab810>
PAIR DELETION
PYTHON SERVER DELETION <grpc._server._Server object at 0x1078ece90> <_cython.cygrpc.Server object at 0x1078ae738>
CANCEL ALL CALLS
NOTIFY SHUTDOWN COMPLETE <_cython.cygrpc.Server object at 0x1078ae738>
PYTHON SERVER DELETION DONE <grpc._server._Server object at 0x1078ece90>
UNIT TEST TEAR DOWN DONE
PAIR DELETION
PYTHON SERVER DELETION <grpc._server._Server object at 0x1078ece90> <_cython.cygrpc.Server object at 0x1078ae738>
PYTHON SERVER DELETION DONE <grpc._server._Server object at 0x1078ece90>
PYTHON SERVER DELETION <grpc._server._Server object at 0x1078ece90> <_cython.cygrpc.Server object at 0x1078ae738>
PYTHON SERVER DELETION DONE <grpc._server._Server object at 0x1078ece90>
PYTHON SERVER STATE DEL
TRUE DEALLOC <_cython.cygrpc.Server object at 0x1078ae738>
>SERVER DESTROIED <_cython.cygrpc.Server object at 0x1078ae738>
>SERVER SHUTDOWN <_cython.cygrpc.Server object at 0x1078ae738>
PYTHON SERVER DELETION <grpc._server._Server object at 0x1078ab810> <_cython.cygrpc.Server object at 0x1078ae0d8>
PYTHON SERVER DELETION DONE <grpc._server._Server object at 0x1078ab810>
PYTHON SERVER STATE DEL
TRUE DEALLOC <_cython.cygrpc.Server object at 0x1078ae0d8>
>SERVER DESTROIED <_cython.cygrpc.Server object at 0x1078ae0d8>
>SERVER SHUTDOWN <_cython.cygrpc.Server object at 0x1078ae0d8>
```

**Comment by ericgribkoff** [comment]
Fixed by #17444
